Add customizable honor code

mylesw27 opened the pull request:
Closes #3087. This PR allows for instructors to customize the honor code that is shown before an exam. This adds a property to the `infoAssessment` JSON schema, `honorCode`, that if set, uses this text rather than the default PL honor code. As discussed [here](https://github.com/PrairieLearn/PrairieLearn/issues/3087#issuecomment-2848094131), we will want to be able to use Markdown and Mustache. I have also included a field in `instructorAssessmentSettings` where this field can be set from in the browser.

reteps commented:
I am curious why we went this direction instead of a dedicated file for the honor code. The experience of editing mustache-formatted HTML embedded inside a JSON file surely cannot be great. For shorter honor codes, this seems fine, but I would envision instructors asking for this kind of feature for longer / more complex honor codes.

jonatanschroeder commented:
> I am curious why we went this direction instead of a dedicated file for the honor code. The experience of editing mustache-formatted HTML embedded inside a JSON file surely cannot be great. For shorter honor codes, this seems fine, but I would envision instructors asking for this kind of feature for longer / more complex honor codes.

Probably related to #4031. Note that there is currently no assessment chunks, so files in the assessment directory (other than clientFilesAssessment) may not be accessible in chunk servers.

nwalters512 commented:
> The experience of editing mustache-formatted HTML embedded inside a JSON file surely cannot be great.

This is why it's being added to the GUI editor for assessment settings; we definitely don't anticipate people editing particularly complex honor codes in JSON. It's also meant to be primarily Markdown, not HTML (though that means that a subset of HTML will be supported).

nwalters512 commented on apps/prairielearn/src/pages/studentAssessment/studentAssessment.ts:
```diff
@@ -79,8 +82,24 @@ router.get(
+    if (res.locals.assessment.type === 'Exam' && res.locals.assessment.require_honor_code) {
+      if (res.locals.assessment.honor_code) {
+        customHonorCode = await markdownToHtml(
+          DOMPurify.sanitize(
```
Can you say more about why you reached for this dependency as opposed to trying to work with the `sanitize` option in the underlying processor?

https://github.com/PrairieLearn/PrairieLearn/blob/c4eaaaad258f00711630c65aedb233bc8a8ebe7c/packages/markdown/src/index.ts#L66

mylesw27 replied:
I originally reached for this dependency based on [this comment](https://github.com/PrairieLearn/PrairieLearn/issues/3087#issuecomment-2848730581) thinking we would want it for sanitizing.  In [59ba1e0](https://github.com/PrairieLearn/PrairieLearn/pull/11953/commits/59ba1e06cb0ccf250972c49f47b264fb102fac1a), I have now removed this dependency and updated the `markdownToHtml` processor to conditionally allow HTML to pass through by updating the usage of `remark2rehype` and `rehypeRaw`.

nwalters512 commented on apps/prairielearn/src/pages/studentAssessment/studentAssessment.ts:
```diff
@@ -79,8 +82,24 @@ router.get(
+        customHonorCode = await markdownToHtml(
+          DOMPurify.sanitize(
+            mustache.render(res.locals.assessment.honor_code, {
+              name: res.locals.user.name,
```
Should we namespace this to `user_name` to make it easier to potentially add more values in the future?

mylesw27 replied:
Yes, updated to `user_name` in [7a0a28e](https://github.com/PrairieLearn/PrairieLearn/pull/11953/commits/7a0a28e9e77e58c914781ab1ea0943da7b638bbe).

nwalters512 replied:
You'll need to update the input's description that still uses `name`.

mylesw27 replied:
Updated the input description to use `{{user_name}}` in [e6c0476](https://github.com/PrairieLearn/PrairieLearn/pull/11953/commits/e6c0476974bf3b136ee627e946f994abb21306f4).

nwalters512 commented on apps/prairielearn/src/schemas/infoAssessment.ts:
```diff
@@ -334,6 +334,10 @@ export const AssessmentJsonSchema = z
       .default(true),
+    honorCode: z
+      .string()
+      .describe('Custom text for the honor code to be accepted before starting exam assessments.')
```
This calls out that it's specifically for exam assessments. Should we update the sync validation code to forbid the use of this property with homeworks?

Maybe we should update the description both here and above to spell this limitation out:

> Custom text for the honor code to be accepted before starting the assessment. Only available for Exam assessments.

mylesw27 replied:
Updated the validation code to forbid the use of `requireHonorCode` or `honorCode` in homeworks in [7a0a28e](https://github.com/PrairieLearn/PrairieLearn/pull/11953/commits/7a0a28e9e77e58c914781ab1ea0943da7b638bbe).

nwalters512 commented on apps/prairielearn/src/pages/instructorAssessmentSettings/instructorAssessmentSettings.ts:
```diff
@@ -206,6 +206,11 @@ router.post(
         );
+        assessmentInfo.honorCode = propertyValueWithDefault(
+          assessmentInfo.honorCode,
+          req.body.honor_code,
```
Consider adding `.trim()` to this.

mylesw27 replied:
Added `.trim()` in [7a0a28e](https://github.com/PrairieLearn/PrairieLearn/pull/11953/commits/7a0a28e9e77e58c914781ab1ea0943da7b638bbe).

nwalters512 commented on packages/markdown/src/index.ts:
```diff
@@ -64,26 +71,30 @@ export function createProcessor({
-      sanitize ? htmlConversion.use(rehypeSanitize) : htmlConversion,
-    )
-    .use(rehypeStringify);
+  const plugins: (Plugin<any, any, any> | PluginTuple<any, any, any>)[] = [
+    remarkParse,
+    remarkMath,
+    ...(mdastVisitors ?? []).map((visitor) => makeHandler(visitor)),
+    makeHandler(visitMathBlock),
+    remarkGfm,
+    [remark2rehype, { allowDangerousHtml: !strictSanitize }],
+    ...(strictSanitize ? [] : [rehypeRaw]),
+    ...(sanitize ? [rehypeSanitize] : []),
+    ...(hastVisitors ?? []).map((visitor) => makeHandler(visitor)),
+    rehypeStringify,
+  ];
+
+  return plugins.reduce((processor: Processor, plugin) => {
+    if (Array.isArray(plugin)) return processor.use(...plugin);
+    return processor.use(plugin);
+  }, unified());
```
@jonatanschroeder tagging you in since you've worked with this Markdown stuff before. I reworked this to make it easier to have conditional plugins in the middle of the chain. I'd appreciate you taking a look at this.

mylesw27 replied:
Updated `strictSanitize` to `allowHtml` and reversed the logic in [e6c0476](https://github.com/PrairieLearn/PrairieLearn/pull/11953/commits/e6c0476974bf3b136ee627e946f994abb21306f4).

nwalters512 commented on packages/markdown/src/index.ts:
```diff
@@ -96,3 +107,16 @@ const inlineProcessor = createProcessor({ hastVisitors: [visitCheckSingleParagra
   return (await (inline ? inlineProcessor : defaultProcessor).process(original)).value.toString();
 }
+
+export async function markdownToHtmlStrict(
+  original: string,
+  { inline }: { inline?: boolean } = {},
+) {
+  return (
+    await (
+      inline
+        ? createProcessor({ hastVisitors: [visitCheckSingleParagraph], strictSanitize: true })
+        : createProcessor({ strictSanitize: true })
+    ).process(original)
+  ).value.toString();
+}
```
Did you consider making `strict` an option to pass to `markdownToHtml`?

mylesw27 replied:
In [e6c0476](https://github.com/PrairieLearn/PrairieLearn/pull/11953/commits/e6c0476974bf3b136ee627e946f994abb21306f4), I renamed `strict` to `allowHtml` for consistency with the processor and added it as an option to be passed to `markdownToHtml` (thus removing `markdownToHtmlStrict`). I also added a `getProcessor` function that memoizes the `createProcessor` call and stores the options as the caching key. I believe I did this correctly, but as I am not too familiar with memoizing, please take a look at that piece of code and make sure it is functioning correctly.

nwalters512 replied:
Looks good! The memoization discussion moved here: https://github.com/PrairieLearn/PrairieLearn/pull/11953#discussion_r2096653633

nwalters512 commented on packages/markdown/src/index.ts:
```diff
@@ -64,26 +71,30 @@ export function createProcessor({
 }: {
   mdastVisitors?: ((ast: MdastRoot) => undefined)[];
   hastVisitors?: ((ast: HastRoot) => undefined)[];
   sanitize?: boolean;
+  strictSanitize?: boolean;
```
What's the difference between `sanitize: true` and `strictSanitize: true`? This could benefit from both documentation and test cases.

mylesw27 replied:
`sanitize: true` is occurring after the Markdown conversion to HTML and is allowing several HTML elements to pass through (see the rehype [defaultSchema](https://github.com/rehypejs/rehype-sanitize?tab=readme-ov-file#defaultschema)). `strictSanitize` has now been renamed `allowHtml` for clarity. If `false`, the original string will be stripped of any html tags during the `remark2rehype` process (effectively removing any HTML tags from the original string). I have added tests for this in [e6c0476](https://github.com/PrairieLearn/PrairieLearn/pull/11953/commits/e6c0476974bf3b136ee627e946f994abb21306f4).

nwalters512 replied:
Nice, I think that's much better!

nwalters512 commented on apps/prairielearn/src/pages/studentAssessment/studentAssessment.ts:
```diff
@@ -79,8 +81,18 @@ router.get(
 
+    let customHonorCode = '';
+    if (res.locals.assessment.type === 'Exam' && res.locals.assessment.require_honor_code) {
+      if (res.locals.assessment.honor_code) {
```
This could be collapsed into the enclosing `if` statement.

mylesw27 replied:
Added this to the enclosing `if` statement in [e6c0476](https://github.com/PrairieLearn/PrairieLearn/pull/11953/commits/e6c0476974bf3b136ee627e946f994abb21306f4).

nwalters512 commented on apps/prairielearn/src/pages/studentAssessment/studentAssessment.html.ts:
```diff
@@ -108,21 +122,30 @@ function StartAssessmentForm({
-      </ul>
-
+      ${customHonorCode
+        ? html`<div class="mx-3 mt-2">${unsafeHtml(customHonorCode)}</div>`
```
Nit, I might do this a little differently:

- Use `px-3 py-2`.
- Use strip the bottom margin from the last child here (you can use the `:last-child` pseudo-class for that)

This will ensure equal vertical spacing above and below the user content.

mylesw27 replied:
Updated to use `px-3 py-2` in [e6c0476](https://github.com/PrairieLearn/PrairieLearn/pull/11953/commits/e6c0476974bf3b136ee627e946f994abb21306f4). I added a stylesheet for this page where I am using `:last-child` to strip the bottom margin here.

nwalters512 replied:
👍 lgtm! You might consider inlining those styles on this page with a `<style>` tag, doing a full network round trip for three lines of CSS seems a little wasteful. We don't have hard and fast rules about this, but I typically only shift styles to an external stylesheet when there are a lot of rules or they're really complex.

nwalters512 reviewed (commented):
I'd like to spend a little more time getting the Markdown stuff polished. Per our discussion in the meeting today, we'd like to work on adding support for rendering `README.md` files in questions soon, which will need the same ability to render sanitized Markdown to HTML.

mylesw27 commented on apps/prairielearn/src/sync/course-db.ts:
```diff
@@ -949,6 +949,10 @@ async function validateAssessment(
+
+    if (assessment.requireHonorCode || assessment.honorCode) {
+      errors.push('Honor code cannot be used for Homework-type assessments');
+    }
```
Split this into two separate, specific checks in [e6c0476](https://github.com/PrairieLearn/PrairieLearn/pull/11953/commits/e6c0476974bf3b136ee627e946f994abb21306f4).

mylesw27 commented on apps/prairielearn/src/pages/instructorAssessmentSettings/instructorAssessmentSettings.html.ts:
```diff
@@ -266,6 +266,31 @@ ${resLocals.assessment.text}</textarea
+                    id="honor_code_group"
+                    ${resLocals.assessment.require_honor_code ? '' : 'hidden'}
+                  >
+                    <label for="honor_code">Custom honor Code</label>
```
Added `form-label` here and to the "text" input in [e6c0476](https://github.com/PrairieLearn/PrairieLearn/pull/11953/commits/e6c0476974bf3b136ee627e946f994abb21306f4).

mylesw27 commented on apps/prairielearn/src/pages/instructorAssessmentSettings/instructorAssessmentSettings.html.ts:
```diff
@@ -266,6 +266,31 @@ ${resLocals.assessment.text}</textarea
+                      ${canEdit ? '' : 'disabled'}
+                    >
+${resLocals.assessment.honor_code}</textarea
+                    >
```
Great idea. I used the same code so I moved this piece to `scripts/lib/resizeTextarea.ts` and updated to call this in both places. I also understand if we don't want to touch `instructorAiGenerateDraftsClient.ts` in this PR and can undo my changes there if needed.

nwalters512 replied:
Nope, making that change is exactly the right move when doing refactoring like this 👍

mylesw27 commented on apps/prairielearn/src/pages/instructorAssessmentSettings/instructorAssessmentSettings.ts:
```diff
@@ -206,6 +206,11 @@ router.post(
         );
+        assessmentInfo.honorCode = propertyValueWithDefault(
+          assessmentInfo.honorCode,
+          req.body.honor_code.trim(),
```
Updated to normalize `\r\n` to `\n` in the honor code in [e6c0476](https://github.com/PrairieLearn/PrairieLearn/pull/11953/commits/e6c0476974bf3b136ee627e946f994abb21306f4).

nwalters512 commented on packages/markdown/src/index.ts:
```diff
@@ -64,35 +71,54 @@ export function createProcessor({
 
-const defaultProcessor = createProcessor();
-const inlineProcessor = createProcessor({ hastVisitors: [visitCheckSingleParagraph] });
+const processorCache = new Map<string, Processor>();
+function getProcessor(options: { inline?: boolean; allowHtml?: boolean } = {}): Processor {
+  const key = JSON.stringify(options);
+  if (!processorCache.has(key)) {
+    processorCache.set(
+      key,
+      createProcessor({
+        hastVisitors: options.inline ? [visitCheckSingleParagraph] : [],
+        allowHtml: options.allowHtml,
+      }),
+    );
+  }
+  return processorCache.get(key) as Processor;
+}
```
@jonatanschroeder do you remember why you memoized initially? Is creating these objects that CPU/memory-intensive?

@mylesw27 if we need to keep memoization, a few notes:

- `JSON.stringify` is sensitive to property order, that is, `{"a":true,"b":true}` is different from `{"b":true,"a":true}`, even though it shouldn't be. Manually constructing a key like `${inline}:${allowHtml}` is probably better.
- `as Processor` works, but we can rewrite this so we don't need casts:

```
let foo = map.get(key);
if (!foo) {
  foo = { thing: 'here' };
  map.set(key, foo);
}
return foo;
```

jonatanschroeder replied:
Good question, I don't know why this was originally memoized. I didn't add the memoization process, the processor was already memoized before I did anything, I just transferred the same idea to the inline processor. I did some digging, and the original implementation of markdown in questions (#1474) already used a memoized processor:

https://github.com/PrairieLearn/PrairieLearn/blob/ca30b2c36eb3497b1eb0c709aa1110d1cc59b6ec/lib/markdown.js#L60-L65

I then added memoized versions of the processor for non-question processing in #6602 to render manual grading feedback, and for inline in #6271 to render rubric grading items. I don't recall at any point testing to check if creating the processor on demand would affect performance.

I did a pretty rudimentary test locally for memoized vs on-demand calls. I converted a very simple string 50000 times with a memoized processor vs on-demand (with an initial call before performance to "dirty" the cache), and I got 12.5-13.5 seconds in the memoized version, and 15-16 seconds creating the processor on demand. I ran the experiment multiple times, and while there was variation between the executions in my machine, memoizing the processor was consistently about 2-3 seconds faster for 50000 runs.

nwalters512 replied:
Ah, I should have looked at the Git history first, apologies. I imagine I did that initially because it was easy (at that point, the processor was static for the lifetime of the process, it didn't change based on any options).

Keeping the memoization certainly won't hurt, I'm fine to keep using it.

mylesw27 replied:
Thank you both for looking into this! I have kept it and updated to use Nathan's suggestions in the first comment above. Now manually constructing the string with `${inline}:${allowHtml}` and updated the logic so we can remove casting.

jonatanschroeder commented on packages/markdown/src/index.ts:
```diff
@@ -64,35 +71,53 @@ export function createProcessor({
-const inlineProcessor = createProcessor({ hastVisitors: [visitCheckSingleParagraph] });
+const processorCache = new Map<string, Processor>();
+function getProcessor(options: { inline?: boolean; allowHtml?: boolean } = {}): Processor {
+  const key = `${options.inline}:${options.allowHtml}`;
```
We probably want `inline: undefined` and `inline: false` to be considered equivalent, right? Should this be:

```suggestion
  const key = `${options.inline ?? false}:${options.allowHtml ?? true}`;
```

nwalters512 replied:
Better yet, we can just make these non-optional, as `markdownToHtml` always provides defaults.

mylesw27 replied:
Made these non-optional in [0b17b42](https://github.com/PrairieLearn/PrairieLearn/pull/11953/commits/0b17b42ccf941335c4a00d1e45af733812ba1fd8).

jonatanschroeder commented:
A note about your use of `@prairielearn/markdown`: this package explicitly handles math content in a way that allows MathJax to take over and draw Math inside it. It also ensures any content inside math blocks is not interpreted as math. So content like `$a _b=c_ d$` is not interpreted as `$a<em>b=c</em>d$`. This is handled in all current uses of the library.

Since the honor code does not currently load/support MathJax, it's probably a good idea to skip the math processor in this specific context. This should allow honor code that includes math delimiters (e.g., dollar sign) not to be treated as math.  Alternatively, you can explicitly support math in the honor code, but I don't think this is necessary or expected to be useful.

mylesw27 commented:
> A note about your use of `@prairielearn/markdown`: this package explicitly handles math content in a way that allows MathJax to take over and draw Math inside it. It also ensures any content inside math blocks is not interpreted as math. So content like `$a _b=c_ d$` is not interpreted as `$a\<em\>b=c\</em\>d$`. This is handled in all current uses of the library.
> 
> Since the honor code does not currently load/support MathJax, it's probably a good idea to skip the math processor in this specific context. This should allow honor code that includes math delimiters (e.g., dollar sign) not to be treated as math. Alternatively, you can explicitly support math in the honor code, but I don't think this is necessary or expected to be useful.

I have added an `allowMath` option that will skip the math processor if `false`. I have added `allowHtml: false` to the use of this package to render the honor code so we are not handling math here.

jonatanschroeder commented on packages/markdown/src/index.test.ts:
```diff
@@ -27,7 +27,7 @@ describe('Markdown processing', () => {
   it('handles inline latex', async () => {
     const question = '$a_1 + a_2 = a_3$';
     const expected = '<p>$a_1 + a_2 = a_3$</p>';
-    await testMarkdown(question, expected);
+    await testMarkdown(question, expected, { allowMath: false });
   });
```
This test is relying on the lack of spaces in the formatting and its interpretation in unified, not on `allowMath`. This should work with the default behaviour, and I don't have a strong opinion on what it should do if `allowMath: false` given its [ambiguity with the spaces and underlines](https://www.markdownguide.org/basic-syntax/#italic-best-practices).

mylesw27 replied:
Oops this was leftover from testing `allowMath` with some of the other tests and I must have missed this one when I was removing them. Removed in  [121e335](https://github.com/PrairieLearn/PrairieLearn/pull/11953/commits/121e33571d597722ba9233dbdce98bc1bfabfe72).

jonatanschroeder commented on packages/markdown/src/index.ts:
```diff
@@ -72,17 +72,19 @@ export function createProcessor({
   const plugins: (Plugin<any, any, any> | PluginTuple<any, any, any>)[] = [
     remarkParse,
-    remarkMath,
+    ...(allowMath ? [remarkMath] : []),
```
Bikeshedding: I'm not 100% sold on the name of the option `allowMath`, it kind of gives the wrong idea. Maybe `renderMath` (through that is not entirely true since this is not actually rendering it) or `tokenizeMath` or `interpretMath`?

mylesw27 replied:
I have updated to `interpretMath` in [127a281](https://github.com/PrairieLearn/PrairieLearn/pull/11953/commits/127a28167e65c420b8d18f8a51c5d4fb32112fda).

jonatanschroeder commented:
Are you planning an integration test here as well, in addition to the markdown test changes? In particular, I think it would be useful to have an assessment in the test course that uses a custom honor code, and have the test check that
1. the assessment with the custom honor code has that particular honor code presented, with the appropriate user name replaced via mustache and some basic Markdown rendered appropriately.
2. possibly some other assessment without the custom honor code shows the default honor code instead.
3. some sanitizable content (e.g., an HTML block that isn't rendered) may be valid as well, but I'm not that concerned here if you test it properly in the package.

I won't block on this, but I think it's a good idea to have it.

mylesw27 commented on packages/markdown/src/index.test.ts:
```diff
@@ -133,4 +133,22 @@ describe('Markdown processing', () => {
+  it('sanitizes HTML if allowHtml is false', async () => {
+    const question = 'testing <h1>html</h1>';
+    const expected = '<p>testing html</p>';
+    await testMarkdown(question, expected, { allowHtml: false });
```
I was not aware of the differences here, thanks! I have updated this test to use an inline tag, `<em>`, instead of `<h1>` for this test. I have also added another test that tests for a full HTML block and that it is fully removed. I added a comment above `allowHtml` noting this behavior.

mylesw27 commented on apps/prairielearn/src/pages/instructorAssessmentSettings/instructorAssessmentSettings.html.ts:
```diff
@@ -266,6 +271,31 @@ ${resLocals.assessment.text}</textarea
+                    >
+                    <small class="form-text text-muted">
+                      Custom honor code text that will be shown to students before starting the
+                      exam. This can use Markdown formatting and a user's name can be included by
```
I have updated the description here noting that this specifically does not support HTML.

jonatanschroeder commented on packages/markdown/src/index.test.ts:
```diff
@@ -133,4 +133,28 @@ describe('Markdown processing', () => {
+  });
+
+  it('sanitizes an HTML block if allowHtml is false', async () => {
+    const question = '<div>testing</div>';
+    const expected = '';
+    await testMarkdown(question, expected, { allowHtml: false });
```
You probably want a test that checks that only the block is removed, not anything else, so it's probably a good idea to have some valid Markdown before and after the block.
```suggestion
    const question = '_Before_ the block\n\n<div>HTML block to be sanitized</div>\n\n**After** the block';
    const expected = '<p><em>Before</em> the block</p>\n<p><strong>After</strong> the block</p>';
    await testMarkdown(question, expected, { allowHtml: false });
```

I'm not 100% sure whether the renderer will put the line break before `</p>`, so you may need to adjust it.

jonatanschroeder commented on packages/markdown/src/index.test.ts:
```diff
@@ -133,4 +133,28 @@ describe('Markdown processing', () => {
+    await testMarkdown(question, expected, { allowHtml: false });
+  });
+
+  it('does not treat math delimiters as math if allowMath is false', async () => {
```
```suggestion
  it('does not treat math delimiters as math if interpretMath is false', async () => {
```

mylesw27 commented:
> Are you planning an integration test here as well, in addition to the markdown test changes? In particular, I think it would be useful to have an assessment in the test course that uses a custom honor code, and have the test check that
> 
> 1. the assessment with the custom honor code has that particular honor code presented, with the appropriate user name replaced via mustache and some basic Markdown rendered appropriately.
> 2. possibly some other assessment without the custom honor code shows the default honor code instead.
> 3. some sanitizable content (e.g., an HTML block that isn't rendered) may be valid as well, but I'm not that concerned here if you test it properly in the package.
> 
> I won't block on this, but I think it's a good idea to have it.

Great suggestion. I have added a custom honor code to exam 2 in the test course and updated the existing `honorCode` test to reflect the new changes. The tests now check that the default honor code is presented if a custom honor code is not supplied. It also checks that if a custom honor code is supplied, it renders correctly including Markdown and mustache.

nwalters512 commented on packages/markdown/src/index.ts:
```diff
@@ -64,35 +71,68 @@ export function createProcessor({
+    allowHtml = true,
+    interpretMath = true,
+  }: { inline?: boolean; allowHtml?: boolean; interpretMath?: boolean } = {},
+) {
+  return (
+    await getProcessor({ inline, allowHtml, interpretMath }).process(original)
+  ).value.toString();
```
Tiny nit, IMO splitting it up makes it slightly easier to read.

```suggestion
) {
  const processor = getProcessor({ inline, allowHtml, interpretMath });
  return (await processor.process(original)).value.toString();
```

nwalters512 reviewed (approved):
I won't block on any of these comments, the core bits here look good!

mylesw27 commented on packages/markdown/src/index.ts:
```diff
@@ -64,35 +71,68 @@ export function createProcessor({
   hastVisitors,
   sanitize = true,
+  // If allowHtml is false, this will remove the tags themselves for inline HTML but will remove the full block for block HTML.
+  // See https://spec.commonmark.org/0.31.2/#raw-html for more details on inline vs block HTML in Markdown.
```
In [f0a3a3c](https://github.com/PrairieLearn/PrairieLearn/pull/11953/commits/f0a3a3c2a7c19586e2d9f7a8ecf8a7c409ac9622), I updated the comment to a multi-line comment and included examples. I also moved it down to type declaration so it shows up in Intellisense.

mylesw27 commented on apps/prairielearn/assets/scripts/instructorAiGenerateDraftsClient.ts:
```diff
@@ -17,31 +17,4 @@ onDocumentReady(() => {
-    observer.unobserve(el);
-  },
-});
+resizeTextArea();
```
Renamed this `autosizeTextareas` in [f0a3a3c](https://github.com/PrairieLearn/PrairieLearn/pull/11953/commits/f0a3a3c2a7c19586e2d9f7a8ecf8a7c409ac9622) and moved this to `behaviors/`. I added it to `application.ts` and removed it from the individual pages that it was included on.
